doc/architecture/005-Message-architecture.md

```diff
@@ -0,0 +1,160 @@
+              region = "us-west-1";
+            };
+          };
+
+          instance = __mkResource {
+            provider = inputs.nixops4-aws.packages.${system}.nixops4-provider;
+            args = ["ec2-instance"];
+            properties = {
+              auth = {
+                # ...
+              };
+              region = "us-west-1";
+              instanceType = "t3.micro";
+              # ...
+            };
+          };
+
+          nixos = __mkResource {
+            provider = inputs.nixops4.packages.${system}.nixops4-local-command;
+            properties = {
+              command = sshWrapper { # akin to https://docs.hercules-ci.com/hercules-ci-effects/reference/nix-functions/ssh
+                destination = "root@${resources.instance.ip}";
+                command = pkgs.writeScript "deploy-nixos" ''
+                  ${(nixpkgs.lib.nixosSystem { modules = [
+                      { services.myapp.settings.bucket = resources.bucket.url; }
+                    ]; }).config.system.toplevel}/bin/apply
+                '';
+              };
+            };
+          };
+
+          dns = __mkResource {
+            provider = inputs.nixops4.packages.${system}.nixops4-aws;
+            args = ["route53-record"];
+            properties = {
+              auth = {
+                # ...
+              };
+              zone = "example.com";
+              name = "myapp.example.com";
+              type = "A";
+              ttl = 300;
+              records = [resources.instance.ip];
+            };
+          };
+        };
+      };
+    };
+}
+```
+
+</details>
+
+## Decision
+
+When the users calls `nix run .#nixops deploy`, the following takes place.
+
+1. Nix builds NixOps4 and runs it.
+2. `nixops4` executable starts its subcommand `deploy`
+3. `deploy` loads the `nixopsProjects.default` expression
+  - Create the evaluator process
+  - Tell the evaluator process to load the `default` project, causing it to evaluate that attribute
+4. `deploy` queries the evaluator for the resources
+  - The evaluator invokes the `nixopsProjects.default` function, passing it the internally defined `__mkResource` function (a Nix "primop"), and a `resources` attribute set containing thunks that refer to unnamed primops.
+  - The evaluator then forces the `resources` attribute set for its names only.
+  - `deploy` receives `["bucket", "dns", "instance", "nixos"]` in the example above. Note that attributes are semantically unordered and returned alphabetically.
+5. `deploy` enters an event loop in which it tries to get as much information about the resources as possible, creates processes for the resources, waits for them to finish and reports the results to evaluator and to the user. We'll describe it using the example project, and in a more linear fashion than a typical execution would be. Where it says "Section", it's a grouping to help with reading, rather than actions that translate into actual code.
```

Comment:
"trying to get as much information about" is related to lazily evaluating, BUT where there are these values that need to spawn I/O and interact with the world to evaluate.

I'm not sure exactly how to put that, but it helps it fit in my brain.

Reply:
Also backtracking is important here!